src/BenchmarkDotNet/Engines/EngineJitStage.cs

```diff
@@ -0,0 +1,246 @@
+        internal Measurement lastMeasurement;
+
+        private readonly int unrollFactor;
+        private readonly Action<long> overheadAction;
+        private readonly Action<long> workloadAction;
+        private readonly IEnumerator<IterationData> enumerator;
+
+        internal EngineFirstJitStage(EngineParameters parameters, int unrollFactor) : base(parameters)
+        {
+            this.unrollFactor = unrollFactor;
+            if (unrollFactor != 1)
+            {
+                overheadAction = parameters.OverheadActionUnroll;
+                workloadAction = parameters.WorkloadActionUnroll;
+            }
+            else
+            {
+                overheadAction = parameters.OverheadActionNoUnroll;
+                workloadAction = parameters.WorkloadActionNoUnroll;
+            }
+            enumerator = EnumerateIterations();
+        }
+
+        internal override List<Measurement> GetMeasurementList() => new(GetMaxMeasurementCount());
+
+        private static int GetMaxMeasurementCount()
+        {
+            int tieredCallCountThreshold = JitInfo.TieredCallCountThreshold;
+            if (JitInfo.IsDPGO)
+            {
+                tieredCallCountThreshold *= 2;
+            }
+            // +1 for first jit, x2 for overhead + workload
+            return (tieredCallCountThreshold + 1) * 2;
+        }
+
+        internal override bool GetShouldRunIteration(List<Measurement> measurements, out IterationData iterationData)
+        {
+            if (measurements.Count > 0)
+            {
+                var measurement = measurements[measurements.Count - 1];
+                if (measurement.IterationMode == IterationMode.Workload)
+                {
+                    lastMeasurement = measurement;
+                }
+            }
+            if (enumerator.MoveNext())
+            {
+                iterationData = enumerator.Current;
+                return true;
+            }
+            enumerator.Dispose();
+            iterationData = default;
+            return false;
+        }
+
+        // We do our best to encourage the jit to fully promote methods to tier1, but tiered jit relies on heuristics,
+        // and we purposefully don't spend too much time in this stage, so we can't guarantee it.
+        // This should succeed for 99%+ of microbenchmarks. For any sufficiently short benchmarks where this fails,
+        // the following stages (Pilot and Warmup) will likely take it the rest of the way. Long-running benchmarks may never fully reach tier1.
+        private IEnumerator<IterationData> EnumerateIterations()
+        {
+            ++iterationIndex;
+            yield return GetDummyIterationData(dummy1Action);
+            yield return GetOverheadNoUnrollIterationData();
+            yield return GetDummyIterationData(dummy2Action);
+            yield return GetWorkloadNoUnrollIterationData();
+            yield return GetDummyIterationData(dummy3Action);
```

Comment:
@AndreyAkinshin You added dummy actions in 2017. I don't know what they are for. Do we still need them?

Reply:
I think it is no longer needed with the new implementation, so I removed them. If you disagree, I can revert that commit @AndreyAkinshin.